File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Cryptopia.java
package com.mobnetic.coinguardian.model.market;

import java.util.List;

import org.json.JSONArray;
import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;

public class Cryptopia extends Market {

	private final static String NAME = "Cryptopia";
	private final static String TTS_NAME = "Cryptopia";
	private final static String URL = "https://www.cryptopia.co.nz/api/GetMarket/%1$s";
	private final static String URL_CURRENCY_PAIRS = "https://www.cryptopia.co.nz/api/GetTradePairs";
	
	public Cryptopia() {
		super(NAME, TTS_NAME, null);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return String.format(URL, checkerInfo.getCurrencyPairId());
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final JSONObject dataJsonObject = jsonObject.getJSONObject("Data");
		ticker.bid = dataJsonObject.getDouble("BidPrice");
		ticker.ask = dataJsonObject.getDouble("AskPrice");
		ticker.vol = dataJsonObject.getDouble("Volume");
		ticker.high = dataJsonObject.getDouble("High");
		ticker.low = dataJsonObject.getDouble("Low");
		ticker.last = dataJsonObject.getDouble("LastPrice");
	}
	
	@Override
	protected String parseErrorFromJsonObject(int requestId, JSONObject jsonObject, CheckerInfo checkerInfo) throws Exception {
		return jsonObject.getString("Message");
	}
	
	// ====================
	// Get currency pairs
	// ====================
	@Override
	public String getCurrencyPairsUrl(int requestId) {
		return URL_CURRENCY_PAIRS;
	}
	
	@Override
	protected void parseCurrencyPairsFromJsonObject(int requestId, JSONObject jsonObject, List<CurrencyPairInfo> pairs) throws Exception {
		final JSONArray dataJsonArray = jsonObject.getJSONArray("Data");
		for(int i=0; i<dataJsonArray.length(); ++i) {
			final JSONObject pairJsonObject = dataJsonArray.getJSONObject(i);
			final String currencyBase = pairJsonObject.getString("Symbol");
			final String currencyCounter = pairJsonObject.getString("BaseSymbol");
			final String pairId = pairJsonObject.getString("Id");
			if (currencyCounter != null && currencyCounter != null && pairId != null) {
				pairs.add(new CurrencyPairInfo(
						currencyBase,
						currencyCounter,
						pairId));
			}
		}
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Cryptopia.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import org.json.JSONObject

class Cryptopia : Market(NAME, TTS_NAME, null) {
    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyPairId)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val dataJsonObject = jsonObject.getJSONObject("Data")
        ticker.bid = dataJsonObject.getDouble("BidPrice")
        ticker.ask = dataJsonObject.getDouble("AskPrice")
        ticker.vol = dataJsonObject.getDouble("Volume")
        ticker.high = dataJsonObject.getDouble("High")
        ticker.low = dataJsonObject.getDouble("Low")
        ticker.last = dataJsonObject.getDouble("LastPrice")
    }

    @Throws(Exception::class)
    override fun parseErrorFromJsonObject(requestId: Int, jsonObject: JSONObject, checkerInfo: CheckerInfo?): String? {
        return jsonObject.getString("Message")
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String? {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairsFromJsonObject(requestId: Int, jsonObject: JSONObject, pairs: MutableList<CurrencyPairInfo>) {
        val dataJsonArray = jsonObject.getJSONArray("Data")
        for (i in 0 until dataJsonArray.length()) {
            val pairJsonObject = dataJsonArray.getJSONObject(i)
            val currencyBase = pairJsonObject.getString("Symbol")
            val currencyCounter = pairJsonObject.getString("BaseSymbol")
            val pairId = pairJsonObject.getString("Id")
            pairs.add(CurrencyPairInfo(
                    currencyBase,
                    currencyCounter,
                    pairId))
        }
    }

    companion object {
        private const val NAME = "Cryptopia"
        private const val TTS_NAME = "Cryptopia"
        private const val URL = "https://www.cryptopia.co.nz/api/GetMarket/%1\$s"
        private const val URL_CURRENCY_PAIRS = "https://www.cryptopia.co.nz/api/GetTradePairs"
    }
}